### PR TITLE
RGB order override for DPI

### PR DESCRIPTION
**Pull request by 6by9**
Quick (untested) mod for https://github.com/raspberrypi/linux/issues/6155

It's either this or adding a load of extra MEDIA_BUS_FMTs.

### CODE REVIEW

**Comment by njhollinghurst** [comment]
Will we be wanting this on RP1 as well?  (BTW the latter can also generate several other pinouts that would leave a gap for I2S, but I expect VC4 can't match those. I mention it in case it helps forward planning...)

File: drivers/gpu/drm/vc4/vc4_dpi.c [hunk 1]
```diff
@@ -205,6 +207,11 @@ static void vc4_dpi_encoder_enable(struct drm_encoder *encoder)
 			}
 		}
 
+		if (dpi->rgb_order_override) {
```
Comment (pelwell) [inline drivers/gpu/drm/vc4/vc4_dpi.c]: Are we happy that it's never necessary to force DPI_ORDER_RGB?

Comment (6by9) [reply drivers/gpu/drm/vc4/vc4_dpi.c]: All the RGB order formats are available via MEDIA_BUS_FMT defines, hence I saw no reason to override it.
I did think about using -1 as "not overridden", but it seemed overkill.

**Comment by 6by9** [comment]
> Will we be wanting this on RP1 as well? (BTW the latter can also generate several other pinouts that would leave a gap for I2S, but I expect VC4 can't match those. I mention it in case it helps forward planning...)

Yes we'll probably want something similar on RP1.
The `dpi` DT node should be remapped to the RP1 peripheral on Pi5 so the property would exist with the overlay change, but the driver would obviously need to look for it. Handling the values in a generic manner might want a little more thought.

File: arch/arm/boot/dts/overlays/README [hunk 2]
```diff
@@ -4988,6 +4988,12 @@ Params: clock-frequency         Display clock frequency (Hz)
                                 Set the default brightness. Normal range 1-16.
                                 (default 16).
         rotate                  Display rotation {0,90,180,270} (default 0)
+        rgb-order               Allow override of RGB order from DPI.
+                                0 = obey MEDIA_BUS_FMT (default)
+                                1 = BGR
+                                2 = GRB
+                                3 = BRG
+
```
Comment (njhollinghurst) [inline arch/arm/boot/dts/overlays/README:4994]: There are 6 permutations including GBR, RBG; I guess those are never wanted?

Comment (6by9) [reply arch/arm/boot/dts/overlays/README:4994]: Those permutations are not available on vc4.

https://github.com/raspberrypi/linux/blob/rpi-6.6.y/drivers/gpu/drm/vc4/vc4_dpi.c#L33-L41

Comment (njhollinghurst) [reply arch/arm/boot/dts/overlays/README:4994]: Ah, fair enough!

Meanwhile I just remembered RP1 would have a problem with all of these for "mode 2" (which really only works when the in-memory fb format is exactly RGB565) but that can't be helped.

**Comment by 6by9** [comment]
Updated and now takes a string for the RGB order so that all orders can be handled in the RP1 driver.

I think this handles the use case fairly cleanly, so dropping the draft status.
(I know I should have updated the DT bindings too, but I doubt this will ever be acceptable to upstream anyway)

File: drivers/gpu/drm/vc4/vc4_dpi.c [hunk 3]
```diff
@@ -361,6 +369,20 @@ static int vc4_dpi_bind(struct device *dev, struct device *master, void *data)
 	if (ret)
 		return ret;
 
+	dpi->rgb_order_override  = -1;
```
Comment (pelwell) [inline drivers/gpu/drm/vc4/vc4_dpi.c]: There's a stray extra space on this line.

Comment (6by9) [reply drivers/gpu/drm/vc4/vc4_dpi.c]: Fixed.

**Pull request merged**